File: web/frontend/src/lib/api/cvs.ts
import type { CvDocuments } from '$lib/stores/documents';
import { apiFetch } from '$lib/api/http';

/**
 * Client for the multi-CV persistence contract (`web/backend/src/
 * rendercv_web/cvs.py`), fixed for Phase 4c:
 *
 * | Method | Path                                    | Notes                          |
 * |--------|-----------------------------------------|---------------------------------|
 * | GET    | /api/cvs                                 | newest first                    |
 * | POST   | /api/cvs                                 | 201, seeded defaults            |
 * | GET    | /api/cvs/{id}                            | 404 if missing/not owned        |
 * | PUT    | /api/cvs/{id}                            | 409 on lost optimistic race     |
 * | POST   | /api/cvs/{id}/duplicate                  | 201 "Copy of {name}"            |
 * | DELETE | /api/cvs/{id}                            | 204                              |
 * | GET    | /api/cvs/{id}/versions                   | newest first                    |
 * | POST   | /api/cvs/{id}/versions/{vid}/restore     | 200, same conflict shape as PUT |
 *
 * Every function accepts an injectable `fetchImpl` (defaulting to the
 * shared, credentialed `apiFetch`) so tests never hit the network -- see
 * `cvs.test.ts`.
 *
 * Why `updated_at` stays a plain string end-to-end: it is an opaque
 * optimistic-concurrency token as far as the client is concerned
 * (`seen_updated_at` round-trips it verbatim back to the server). Parsing it
 * into a `Date` and re-serializing would risk losing precision the server
 * didn't -- comparisons and display both work fine off the ISO string.
 */

export interface CvSummary {
	id: number;
	name: string;
	updatedAt: string;
}

export interface CvDetail extends CvSummary {
	documents: CvDocuments;
}

export interface CvVersionSummary {
	id: number;
	createdAt: string;
}

export type UpdateCvResult =
	| { ok: true; updatedAt: string }
	| { ok: false; kind: 'not-found' }
	| { ok: false; kind: 'too-large' }
	| { ok: false; kind: 'conflict'; current: { updatedAt: string; documents: CvDocuments } }
	| { ok: false; kind: 'error'; message: string };

interface CvDocumentsPayload {
	cv_yaml: string;
	design_yaml: string;
	locale_yaml: string;
	settings_yaml: string;
}

interface CvSummaryPayload {
	id: number;
	name: string;
	updated_at: string;
}

interface CvDetailPayload extends CvSummaryPayload {
	documents: CvDocumentsPayload;
}

interface CvVersionSummaryPayload {
	id: number;
	created_at: string;
}

/** Converts the wire shape (`cv_yaml`, ...) into the app's `CvDocuments` (`cv`, ...). */
export function toDocuments(payload: CvDocumentsPayload): CvDocuments {
	return {
		cv: payload.cv_yaml,
		design: payload.design_yaml,
		locale: payload.locale_yaml,
		settings: payload.settings_yaml
	};
}

/** Converts `CvDocuments` (`cv`, ...) into the wire shape (`cv_yaml`, ...). */
export function toPayload(docs: CvDocuments): CvDocumentsPayload {
	return {
		cv_yaml: docs.cv,
		design_yaml: docs.design,
		locale_yaml: docs.locale,
		settings_yaml: docs.settings
	};
}

function toSummary(payload: CvSummaryPayload): CvSummary {
	return { id: payload.id, name: payload.name, updatedAt: payload.updated_at };
}

function toDetail(payload: CvDetailPayload): CvDetail {
	return { ...toSummary(payload), documents: toDocuments(payload.documents) };
}

async function unexpectedError(response: Response): Promise<{ ok: false; kind: 'error'; message: string }> {
	const text = await response.text().catch(() => '');
	return { ok: false, kind: 'error', message: text || `Request failed with status ${response.status}` };
}

/** `GET /api/cvs` -- every saved CV, newest-updated first. */
export async function listCvs(fetchImpl: typeof fetch = apiFetch): Promise<CvSummary[]> {
	const response = await fetchImpl('/api/cvs');
	if (!response.ok) throw new Error(`GET /api/cvs failed with status ${response.status}`);
	const body = (await response.json()) as CvSummaryPayload[];
	return body.map(toSummary);
}

/** `POST /api/cvs` -- creates a new CV seeded with the editor's default documents. */
export async function createCv(name?: string, fetchImpl: typeof fetch = apiFetch): Promise<CvDetail> {
	const response = await fetchImpl('/api/cvs', {
		method: 'POST',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify(name ? { name } : {})
	});
	if (!response.ok) throw new Error(`POST /api/cvs failed with status ${response.status}`);
	return toDetail((await response.json()) as CvDetailPayload);
}

/** `GET /api/cvs/{id}` -- one CV, in full. `null` if it doesn't exist (or isn't owned). */
export async function getCv(id: number, fetchImpl: typeof fetch = apiFetch): Promise<CvDetail | null> {
	const response = await fetchImpl(`/api/cvs/${id}`);
	if (response.status === 404) return null;
	if (!response.ok) throw new Error(`GET /api/cvs/${id} failed with status ${response.status}`);
	return toDetail((await response.json()) as CvDetailPayload);
}

export interface UpdateCvRequest {
	name: string;
	documents: CvDocuments;
	seenUpdatedAt: string;
}

/** `PUT /api/cvs/{id}` -- the autosave write; fails with `conflict` if the CV changed since `seenUpdatedAt`. */
export async function updateCv(
	id: number,
	request: UpdateCvRequest,
	fetchImpl: typeof fetch = apiFetch,
	init: RequestInit = {}
): Promise<UpdateCvResult> {
	const response = await fetchImpl(`/api/cvs/${id}`, {
		...init,
		method: 'PUT',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify({
			name: request.name,
			documents: toPayload(request.documents),
			seen_updated_at: request.seenUpdatedAt
		})
	});

	if (response.status === 200) {
		const body = (await response.json()) as { updated_at: string };
		return { ok: true, updatedAt: body.updated_at };
	}
	if (response.status === 404) return { ok: false, kind: 'not-found' };
	if (response.status === 413) return { ok: false, kind: 'too-large' };
	if (response.status === 409) {
		const body = (await response.json()) as { current: { updated_at: string; documents: CvDocumentsPayload } };
		return {
			ok: false,
			kind: 'conflict',
			current: { updatedAt: body.current.updated_at, documents: toDocuments(body.current.documents) }
		};
	}
	return await unexpectedError(response);
}

/** `POST /api/cvs/{id}/duplicate` -- creates a copy named `Copy of {name}`. */
export async function duplicateCv(id: number, fetchImpl: typeof fetch = apiFetch): Promise<CvDetail | null> {
	const response = await fetchImpl(`/api/cvs/${id}/duplicate`, { method: 'POST' });
	if (response.status === 404) return null;
	if (!response.ok) throw new Error(`POST /api/cvs/${id}/duplicate failed with status ${response.status}`);
	return toDetail((await response.json()) as CvDetailPayload);
}

/** `DELETE /api/cvs/{id}`. Returns `false` if the CV didn't exist (or wasn't owned). */
export async function deleteCv(id: number, fetchImpl: typeof fetch = apiFetch): Promise<boolean> {
	const response = await fetchImpl(`/api/cvs/${id}`, { method: 'DELETE' });
	if (response.status === 404) return false;
	if (response.status !== 204) throw new Error(`DELETE /api/cvs/${id} failed with status ${response.status}`);
	return true;
}

/** `GET /api/cvs/{id}/versions` -- autosave snapshots, newest first. */
export async function listVersions(
	id: number,
	fetchImpl: typeof fetch = apiFetch
): Promise<CvVersionSummary[] | null> {
	const response = await fetchImpl(`/api/cvs/${id}/versions`);
	if (response.status === 404) return null;
	if (!response.ok) throw new Error(`GET /api/cvs/${id}/versions failed with status ${response.status}`);
	const body = (await response.json()) as CvVersionSummaryPayload[];
	return body.map((v) => ({ id: v.id, createdAt: v.created_at }));
}

/** `POST /api/cvs/{id}/versions/{versionId}/restore` -- restores a snapshot as a new autosave write. */
export async function restoreVersion(
	id: number,
	versionId: number,
	fetchImpl: typeof fetch = apiFetch
): Promise<UpdateCvResult> {
	const response = await fetchImpl(`/api/cvs/${id}/versions/${versionId}/restore`, { method: 'POST' });

	if (response.status === 200) {
		const body = (await response.json()) as { updated_at: string };
		return { ok: true, updatedAt: body.updated_at };
	}
	if (response.status === 404) return { ok: false, kind: 'not-found' };
	if (response.status === 409) {
		const body = (await response.json()) as { current: { updated_at: string; documents: CvDocumentsPayload } };
		return {
			ok: false,
			kind: 'conflict',
			current: { updatedAt: body.current.updated_at, documents: toDocuments(body.current.documents) }
		};
	}
	return await unexpectedError(response);
}

File: web/backend/src/rendercv_web/cvs.py
"""Multi-CV persistence: `/api/cvs` and its version-history sub-resource.

Why:
    Phase 4b (docs/plans/completed/cv-editor-web-app.md): the sidebar's list of
    saved CVs, the autosave write, and undo history, all scoped to the
    anonymous session identified by `auth.CurrentUser`. Every read and
    write is scoped to `current_user.id`; a CV id that exists but belongs to
    someone else is treated exactly like a CV id that doesn't exist at all
    (guardrail: never leak ownership via a 403).
"""

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from .auth import CurrentUser
from .db import repository
from .db.models import Cv
from .db.session import get_session
from .defaults import (
    DEFAULT_CV_NAME,
    DEFAULT_CV_YAML,
    DEFAULT_DESIGN_YAML,
    DEFAULT_LOCALE_YAML,
    DEFAULT_SETTINGS_YAML,
)
from .errors import CvConflictError
from .limits import enforce_documents_size_cap
from .models import (
    MAX_CV_NAME_LENGTH,
    CvConflictCurrent,
    CvCreateRequest,
    CvDetail,
    CvDocumentsPayload,
    CvSummary,
    CvUpdateRequest,
    CvUpdateResponse,
    CvVersionSummary,
)

router = APIRouter(tags=["cvs"])

SessionDep = Annotated[Session, Depends(get_session)]

VERSIONS_TO_KEEP = 50


def cv_documents(cv: Cv) -> CvDocumentsPayload:
    """Extract a CV row's four YAML documents into the API payload shape.

    Args:
        cv: The database row.

    Returns:
        The four documents as a `CvDocumentsPayload`.
    """
    return CvDocumentsPayload(
        cv_yaml=cv.cv_yaml,
        design_yaml=cv.design_yaml,
        locale_yaml=cv.locale_yaml,
        settings_yaml=cv.settings_yaml,
    )


def cv_to_summary(cv: Cv) -> CvSummary:
    """Convert a CV row into a `GET /api/cvs` list entry.

    Args:
        cv: The database row.

    Returns:
        The row's id, name, and `updated_at`.
    """
    return CvSummary(id=cv.id, name=cv.name, updated_at=cv.updated_at)


def cv_to_detail(cv: Cv) -> CvDetail:
    """Convert a CV row into the full `CvDetail` response shape.

    Args:
        cv: The database row.

    Returns:
        The row's id, name, `updated_at`, and its four documents.
    """
    return CvDetail(
        id=cv.id, name=cv.name, updated_at=cv.updated_at, documents=cv_documents(cv)
    )


def get_owned_cv_or_404(session: Session, cv_id: int, user_id: int) -> Cv:
    """Fetch a CV owned by `user_id`, or raise the API's not-found error.

    Args:
        session: The database session.
        cv_id: The CV to fetch.
        user_id: The caller's user id.

    Returns:
        The owned `Cv` row.

    Raises:
        HTTPException: 404 if no such CV is owned by `user_id` -- this is
            also what a CV owned by someone else looks like, deliberately.
    """
    cv = repository.get_cv(session, cv_id, user_id)
    if cv is None:
        raise HTTPException(status_code=404, detail="CV not found.")
    return cv


def apply_update_result(
    session: Session, cv_id: int, user_id: int, result: repository.CvUpdateResult
) -> CvUpdateResponse:
    """Turn an `update_cv_conditional` outcome into a response, or raise.

    Why:
        Shared by the autosave write and version restore, both of which
        must record a version snapshot and prune old ones on success, and
        surface the same 404/409 shapes on failure (docs/plans/active/
        cv-editor-web-app.md, Phase 4).

    Args:
        session: The database session.
        cv_id: The CV that was written to.
        user_id: The caller's user id.
        result: The outcome of `repository.update_cv_conditional`.

    Returns:
        The new `updated_at` on success.

    Raises:
        HTTPException: 404 if the CV doesn't exist or isn't owned.
        CvConflictError: 409 (translated by the exception boundary) if the
            write lost the optimistic-concurrency race; carries the
            server's current state so the client can reconcile.
    """
    if result.conflict == "not_found":
        raise HTTPException(status_code=404, detail="CV not found.")

    if result.conflict == "stale":
        current = get_owned_cv_or_404(session, cv_id, user_id)
        raise CvConflictError(
            current=CvConflictCurrent(
                updated_at=current.updated_at, documents=cv_documents(current)
            )
        )

    cv = result.cv
    if cv is None:  # pragma: no cover - conflict is None iff cv is set
        raise HTTPException(status_code=404, detail="CV not found.")

    repository.add_version(
        session, cv_id, cv.cv_yaml, cv.design_yaml, cv.locale_yaml, cv.settings_yaml
    )
    repository.prune_versions(session, cv_id, keep=VERSIONS_TO_KEEP)
    return CvUpdateResponse(updated_at=cv.updated_at)


@router.get("/api/cvs", response_model=list[CvSummary])
def list_cvs(current_user: CurrentUser, session: SessionDep) -> list[CvSummary]:
    """List the current session's CVs, most recently updated first.

    Args:
        current_user: The session's `User`, resolved from the session
            cookie.
        session: The database session.

    Returns:
        One `CvSummary` per owned CV, newest-updated first.
    """
    cvs = repository.list_cvs(session, current_user.id)
    return [cv_to_summary(cv) for cv in cvs]


@router.post("/api/cvs", response_model=CvDetail, status_code=201)
def create_cv(
    request: CvCreateRequest, current_user: CurrentUser, session: SessionDep
) -> CvDetail:
    """Create a new CV seeded with the editor's default documents.

    Why:
        Seeds the same placeholder content the frontend shows for a
        brand-new, unsaved session (`defaults.py`), so saving it
        immediately is a no-op from the user's point of view.

    Args:
        request: The optional display name for the new CV.
        current_user: The session's `User`, resolved from the session
            cookie.
        session: The database session.

    Returns:
        The newly created CV, in full.
    """
    cv = repository.create_cv(
        session,
        current_user.id,
        name=request.name or DEFAULT_CV_NAME,
        cv_yaml=DEFAULT_CV_YAML,
        design_yaml=DEFAULT_DESIGN_YAML,
        locale_yaml=DEFAULT_LOCALE_YAML,
        settings_yaml=DEFAULT_SETTINGS_YAML,
    )
    return cv_to_detail(cv)


@router.get("/api/cvs/{cv_id}", response_model=CvDetail)
def get_cv(cv_id: int, current_user: CurrentUser, session: SessionDep) -> CvDetail:
    """Fetch one CV, in full.

    Args:
        cv_id: The CV to fetch.
        current_user: The session's `User`, resolved from the session
            cookie.
        session: The database session.

    Returns:
        The CV's id, name, `updated_at`, and documents.

    Raises:
        HTTPException: 404 if the CV doesn't exist or isn't owned by the
            current session.
    """
    cv = get_owned_cv_or_404(session, cv_id, current_user.id)
    return cv_to_detail(cv)


@router.put("/api/cvs/{cv_id}", response_model=CvUpdateResponse)
def update_cv(
    cv_id: int,
    request: CvUpdateRequest,
    current_user: CurrentUser,
    session: SessionDep,
) -> CvUpdateResponse:
    """Autosave a CV, iff it hasn't changed since the client last read it.

    Args:
        cv_id: The CV to update.
        request: The new name, documents, and the `updated_at` the client
            last saw.
        current_user: The session's `User`, resolved from the session
            cookie.
        session: The database session.

    Returns:
        The new `updated_at` on success.

    Raises:
        HTTPException: 404 if the CV doesn't exist or isn't owned, 413 if
            any document exceeds the size cap.
        CvConflictError: 409 (translated by the exception boundary) if the
            write lost the optimistic-concurrency race.
    """
    enforce_documents_size_cap(request.documents.model_dump())
    result = repository.update_cv_conditional(
        session,
        cv_id,
        current_user.id,
        request.seen_updated_at,
        name=request.name,
        cv_yaml=request.documents.cv_yaml,
        design_yaml=request.documents.design_yaml,
        locale_yaml=request.documents.locale_yaml,
        settings_yaml=request.documents.settings_yaml,
    )
    return apply_update_result(session, cv_id, current_user.id, result)


@router.post("/api/cvs/{cv_id}/duplicate", response_model=CvDetail, status_code=201)
def duplicate_cv(
    cv_id: int, current_user: CurrentUser, session: SessionDep
) -> CvDetail:
    """Create a copy of a CV, named `Copy of {name}`.

    Args:
        cv_id: The CV to duplicate.
        current_user: The session's `User`, resolved from the session
            cookie.
        session: The database session.

    Returns:
        The newly created copy, in full.

    Raises:
        HTTPException: 404 if the source CV doesn't exist or isn't owned.
    """
    source = get_owned_cv_or_404(session, cv_id, current_user.id)
    copy = repository.create_cv(
        session,
        current_user.id,
        name=f"Copy of {source.name}"[:MAX_CV_NAME_LENGTH],
        cv_yaml=source.cv_yaml,
        design_yaml=source.design_yaml,
        locale_yaml=source.locale_yaml,
        settings_yaml=source.settings_yaml,
    )
    return cv_to_detail(copy)


@router.delete("/api/cvs/{cv_id}", status_code=204)
def delete_cv(cv_id: int, current_user: CurrentUser, session: SessionDep) -> None:
    """Delete a CV and its version history.

    Why:
        Returns `None` rather than a fresh `Response` object: see the
        identical note on `preferences.set_preference` -- returning a new
        `Response` here would silently drop any session cookie the auth
        dependency just set on the shared response object.

    Args:
        cv_id: The CV to delete.
        current_user: The session's `User`, resolved from the session
            cookie.
        session: The database session.

    Raises:
        HTTPException: 404 if the CV doesn't exist or isn't owned.
    """
    deleted = repository.delete_cv(session, cv_id, current_user.id)
    if not deleted:
        raise HTTPException(status_code=404, detail="CV not found.")


@router.get("/api/cvs/{cv_id}/versions", response_model=list[CvVersionSummary])
def list_versions(
    cv_id: int, current_user: CurrentUser, session: SessionDep
) -> list[CvVersionSummary]:
    """List a CV's autosave snapshots, newest first.

    Args:
        cv_id: The CV whose versions to list.
        current_user: The session's `User`, resolved from the session
            cookie.
        session: The database session.

    Returns:
        One `CvVersionSummary` per snapshot, newest first.

    Raises:
        HTTPException: 404 if the CV doesn't exist or isn't owned.
    """
    versions = repository.list_versions(session, cv_id, current_user.id)
    if versions is None:
        raise HTTPException(status_code=404, detail="CV not found.")
    return [
        CvVersionSummary(id=version.id, created_at=version.created_at)
        for version in versions
    ]


@router.post(
    "/api/cvs/{cv_id}/versions/{version_id}/restore", response_model=CvUpdateResponse
)
def restore_version(
    cv_id: int,
    version_id: int,
    current_user: CurrentUser,
    session: SessionDep,
) -> CvUpdateResponse:
    """Restore a version snapshot as a new autosave write.

    Why:
        Applied as a normal conditional update (not an in-place rewrite) so
        it bumps `updated_at`, records its own new version, and can lose an
        optimistic-concurrency race exactly like any other autosave would
        (docs/plans/completed/cv-editor-web-app.md, Phase 4).

    Args:
        cv_id: The CV to restore a version onto.
        version_id: The version snapshot to restore.
        current_user: The session's `User`, resolved from the session
            cookie.
        session: The database session.

    Returns:
        The new `updated_at` on success.

    Raises:
        HTTPException: 404 if the CV or the version doesn't exist or isn't
            owned.
        CvConflictError: 409 (translated by the exception boundary) if the
            write lost the optimistic-concurrency race against a
            concurrent autosave.
    """
    cv = get_owned_cv_or_404(session, cv_id, current_user.id)
    version = repository.get_version(session, cv_id, current_user.id, version_id)
    if version is None:
        raise HTTPException(status_code=404, detail="Version not found.")

    result = repository.update_cv_conditional(
        session,
        cv_id,
        current_user.id,
        cv.updated_at,
        name=cv.name,
        cv_yaml=version.cv_yaml,
        design_yaml=version.design_yaml,
        locale_yaml=version.locale_yaml,
        settings_yaml=version.settings_yaml,
    )
    return apply_update_result(session, cv_id, current_user.id, result)
